### DISCUSSION TITLE
Getting invalid "function $FN cannot return external type from dependency module $TYPE "

### DISCUSSION THREAD

**Discussion by marcosnils** [dagger/dagger#12191] [Help]
# Problem

Getting the following error when trying to define a type with the same  name a as a module type:

```
Error: input: module.withSource.initialize resolve: failed to initialize module: failed to add object to module "program-backend": failed to validate type def: object "ProgramBackend" function "Cache" cannot return external type from dependency module "cache"
```

main module: 

```
package main

type ProgramBackend struct{}

type Cache struct{}

// Returns a container that echoes whatever string argument is provided
func (m *ProgramBackend) Cache(stringArg string) *Cache {
	dag.Cache().Check()
	return &Cache{}
}
```

dependent module: 

```
package main

import (
	"dagger/cache/internal/dagger"
)

type Cache struct{}

// Returns a container that echoes whatever string argument is provided
func (m *Cache) Check() *dagger.Container {
	return dag.Container().From("alpine:latest").WithExec([]string{"echo", "hello"})
}
```

as shown above, given that the `ProgramBackend's` is not effectively returning the dependent's module type but its own, I was expecting this to work.

**Comment by marcosnils** [comment]
cc @helderco might know what's happening here

**Comment by jedevc** [comment]
Hm, this looks like a naming confusion error :thinking: 

I suspect we're getting something wrong in the go codegen / engine here - this is where the error is from: https://github.com/dagger/dagger/blob/8b3b8af59328d341a0486eaf1463736ca376849b/core/module.go#L475-L484

(`already validated, skip` is *not* the right comment for this code :thinking:)

@marcosnils can you confirm, the `main` module has your `dependent` module installed? Also, what are the names of the modules? If the dependent module is called `Cache` this is probably the issue.

I think what's happening is:
- All dependency modules have one top level type - e.g. if I import a module called `cache`, the top-level type is `Cache` - this is part of my served schema.
- I declare a new type called `Cache` - this doesn't conflict at the go-level, since the top-level type is in the `internal/dagger` package as `dagger.Cache`.
- But, in the generated code for `ProgramBackend.Cache`, I declare the return type as `"Cache"` - this name is now ambiguous, does it refer to the type I just generated, or does it refer to the top-level type from my dependency? And we're getting the wrong one.
- For now, it makes sense to just switch this logic - because of the ordering of the linked code, we prefer dependencies, since you can't return dependencies, that doesn't make sense for now.
	- Currently, we determine if a type is local to a module if it isn't part of a dependency - this logic is flawed.
	- We could update to try and look at `ObjectDefs` to see if it's local? And only then look at dependencies? This is neat, but tricky to do, since `namespaceObject` occurs before we actually add to `ObjectDefs`. Also doesn't work for circular definitions - e.g. `a.func` returns a `b` and `b.func` returns an `a`.
	- Maybe the SDK should flag use of a dependency type vs a local type, instead of relying on one `WithObject`?

That *should* fix the issue? That said, I wouldn't be surprised if we saw other surprising behavior as part of this case, it's probably worth temporarily hacking the check to get the right object to see if it's actually possible to enable this behavior.
